frappe/database/database.py

```diff
@@ -264,21 +276,50 @@ def sql(
 		if not self._cursor.description:
 			return ()
 
-		self.last_result = self._transform_result(self._cursor.fetchall())
+		last_result = self._transform_result(self._cursor.fetchall())
+		if as_iterator:
+			return self._return_as_iterator(
+				last_result, pluck=pluck, as_dict=as_dict, as_list=as_list, update=update
+			)
 
 		if pluck:
-			return [r[0] for r in self.last_result]
+			last_result = [r[0] for r in last_result]
+			self._clean_up()
+			return last_result
 
 		# scrub output if required
 		if as_dict:
-			ret = self.fetch_as_dict()
+			last_result = self.fetch_as_dict(last_result)
 			if update:
-				for r in ret:
+				for r in last_result:
 					r.update(update)
-			return ret
+
 		elif as_list:
-			return self.convert_to_lists(self.last_result)
-		return self.last_result
+			last_result = self.convert_to_lists(last_result)
+
+		self._clean_up()
+		return last_result
+
+	def _return_as_iterator(self, result, *, pluck, as_dict, as_list, update):
+		if pluck:
+			for row in result:
+				yield row[0]
+
+		elif as_dict:
+			keys = [column[0] for column in self._cursor.description]
+			for row in result:
+				row = frappe._dict(zip(keys, row))
+				if update:
+					row.update(update)
+				yield row
+
+		elif as_list:
+			for row in result:
+				yield list(row)
+		else:
+			yield from result
```

Comment:
This is actually useless, maybe shouldn't allow using iterator if not list/dict.